rfc/rfc-76/rfc-76.md

```diff
@@ -0,0 +1,143 @@
+2. report RO table data completion and freshness.
+
+With the data completion and freshness of RO table, we can expand use-case of RO table.
+That is given event time threshold T, the log files before T can be compacted, then resulting RO table obtains all data before T, with low query latency.
+
+Currently, we introduced new file slicing algorithm based on barrier and generating compaction plan based on completion time [[HUDI-6495][RFC-66]](https://github.com/apache/hudi/pull/7907).
+
+With the barrier file slicing, file slices are split by the barriers extracted from every instant time of base files 
+or the first instant of log file if no base file exists.
+
+When generating compaction plan, the scheduler will filter all log files with completion time before the `compactionInstant`. 
+And the compaction operation will generate a new base file with `compactionInstant` as its instant time;
+
+So the file slices are split naturally by the `compactionInstant` for MOR cases. 
+
+So we can replace the `compactionInstant` with `eventTimeInstant` computed by `EventTimeBasedCompactionStrategy` when generating plan. 
+In this case, file slices are split by `eventTimeInstant` instead of `compactionInstant`. 
+This is the base for `EventTimeBasedCompactionStrategy`.
+
+In .commit/.deltacommit metadata, we can extract the min/max event time for each base file or log file and take the min/max event time as data completion and freshness for that base file or log file. 
+This is the base for data completion and freshness of RO table.
+
+## Implementation
+
+With the ability of new file slice algorithm and compaction based on completion time, we can choose to
+
+### Schedule Compaction Plan
+The `EventTimeBasedCompactionStrategy` strategy is responsible to calculate the event time threshold for compaction plan.
+Implement a method to calculate the event time threshold:
+```java
+    // return the threshold as instant time format
+    public String calcEventTimeThreshold(String compactionInstant, HoodieWriteConfig config)
+```
+
+When generating compaction plan, we take the event time threshold as the next base instant time and save it to `CompactionOpertion::nextBaseInstantTime`.
+
+
+### Execute Compaction Plan
+Get `nextBaseInstantTime` from `CompactionOpertion`, write base file to the `nextBaseInstantTime`.
+
+Sample code in HoodieCompactor:
+```java
+    String nextBaseInstantTime = compactionOpertion.getNextBaseInstantTime();
+    result = executionHelper.writeFileAndGetWriteStats(compactionHandler, operation, nextBaseInstantTime, scanner, oldDataFileOpt);
+```
+
+Follow up:
+
+With the `EventTimeBasedCompactionStrategy` above, there can be data with event time after the threshold being compacted.
+For cases users want exact event time split (users don't want any data after threshold exist in RO table),
+we need to perform the log record filter in compaction execution based on the time threshold.
```

Comment:
> users don't want any data after threshold exist in RO table

What threshold are you refering to?

Reply:
The threshold refers to the generated `eventTimeInstant`.
For example, there may be cases that we make daily snapshots from hudi table, yesterday's snapshot only contains data before yesterday. In this case, the 0 o'clock is what I mean threshold here.

Reply:
The compactor can ensure this if we if there is no clock skew, in our Hudi implementation, we will by default wait another 200ms each time generating a new timstamp. But in general, I don't think we can solve the problem of data drift from late arrived records.

let's say we have a thresold of t3, and we have another 2 records arrived at t2 and t4, but their real event time is t1, how do you handle these 2 records.

Reply:
Yeah, generally, we can not solve data drift issue from late records simply. 

But given a precondition that the data delay wont exceed `late_interval` (eg. 10 mins), then we can schedule the compaction at timestamp `t3 + late_inteval` with eventTimeInstant `t3`. With the precondition, all data with real event time before `t3` should be ingested.

Reply:
> then we can schedule the compaction at timestamp t3 + late_inteval with eventTimeInstant t3.

The additional wait for time interval `late_inteval` does not make sense because our instant start/completion are proctime insttead of event time. The result for scheduling at time `t3 + late_inteval` and `t3` should be the same.

Reply:
I see, will remove this part.